src/Sylius/Bundle/CoreBundle/Doctrine/ORM/ProductRepository.php

```diff
@@ -29,13 +29,17 @@ class ProductRepository extends BaseProductRepository implements ProductReposito
      */
     public function createByTaxonPaginator(TaxonInterface $taxon, array $criteria = [])
     {
+        $root = $taxon->isRoot() ? $taxon : $taxon->getRoot();
+
         $queryBuilder = $this->createQueryBuilder('o');
         $queryBuilder
             ->innerJoin('o.taxons', 'taxon')
+            ->andWhere($queryBuilder->expr()->eq('o.root', ':root'))
```

Comment:
If parent's left is always smaller than child's left and parent's right is always bigger than child's left, there is no way that this issue may exist (assuming that nested set tree implementation is correct). 

Which version do you use? I remember that there were some issues when we used both nested set tree and softdeletable back in the days. (This may still exist in newer versions if using the same database then and now).